lib/schedule/data.ex

```diff
@@ -214,87 +266,182 @@ defmodule Schedule.Data do
+  def run_for_trip(%{runs: runs_table} = tables, run_id, trip_id) do
+    trip = trip(tables, trip_id)
 
-    cond do
-      trip != nil and run_id != nil and trip.schedule_id != nil ->
-        runs[{trip.schedule_id, run_id}]
+    run_key =
+      cond do
+        trip != nil and run_id != nil and trip.schedule_id != nil ->
+          {trip.schedule_id, run_id}
 
-      trip != nil and trip.schedule_id != nil ->
-        runs[{trip.schedule_id, trip.run_id}]
+        trip != nil and trip.schedule_id != nil ->
+          {trip.schedule_id, trip.run_id}
 
-      true ->
-        nil
+        true ->
+          nil
+      end
+
+    if is_nil(run_key) do
+      nil
+    else
+      case :mnesia.dirty_read(runs_table, run_key) do
+        [{^runs_table, _, run}] -> run
+        _ -> nil
+      end
     end
   end
 
-  @spec block_for_trip(t(), Schedule.Trip.id()) :: Block.t() | nil
+  @spec block_for_trip(tables(), Schedule.Trip.id()) :: Block.t() | nil
   def block_for_trip(
-        %__MODULE__{
-          trips: trips,
-          blocks: blocks
-        },
+        %{blocks: blocks_table} = tables,
         trip_id
       ) do
-    trip = trips[trip_id]
+    trip = trip(tables, trip_id)
 
     if trip != nil && trip.schedule_id != nil do
       # we have HASTUS data for this trip
-      blocks[{trip.schedule_id, trip.block_id}]
+      case :mnesia.dirty_read(blocks_table, {trip.schedule_id, trip.block_id}) do
+        [{^blocks_table, _, block}] -> block
+        _ -> nil
+      end
     else
       nil
     end
   end
 
-  @spec swings_for_route(t(), Route.id(), Util.Time.timestamp(), Util.Time.timestamp()) ::
+  @spec swings_for_route(tables(), Route.id(), Util.Time.timestamp(), Util.Time.timestamp()) ::
           [Swing.t()] | nil
   def swings_for_route(
-        %__MODULE__{calendar: calendar, swings: swings},
+        %{swings: swings_table} = tables,
         route_id,
         start_time,
         end_time
       ) do
     dates = potentially_active_service_dates(start_time, end_time)
-    active_services = Map.take(calendar, dates)
 
-    Enum.flat_map(active_services, fn {_data, service_ids} ->
+    dates
+    |> active_services_on_dates(tables)
+    |> Enum.flat_map(fn {_data, service_ids} ->
       Enum.flat_map(service_ids, fn service_id ->
-        Map.get(swings, {service_id, route_id}, [])
+        case :mnesia.dirty_match_object({swings_table, {service_id, route_id}, :_}) do
+          swing_records when is_list(swing_records) -> Enum.map(swing_records, &elem(&1, 2))
+          _ -> []
+        end
       end)
     end)
   end
 
   # Initialization
 
+  @spec initialize_tables(tables()) :: :ok
+  def initialize_tables(tables) do
+    Enum.each(@table_schema, fn {table_key, columns, extra_indices} ->
+      # TODO: drop table if it already exists before attempting to create
+      {:atomic, :ok} =
+        :mnesia.create_table(tables[table_key],
+          attributes: columns,
+          index: extra_indices,
```

Comment:
question: I'm not sure I saw any queries using the extra indicies?

Reply:
Hmm, so the [documentation](https://www.erlang.org/doc/man/mnesia.html#match_object-3) on `match_object` makes it sound like it uses "indexes" (plural) if available, but then says "[h]owever, no heuristics are performed to select the best index" (single), so it's not clear to me on reflection if it uses the extra indexes automatically.

The instances in which I _intend_ to use the extra indexes are functions like `active_trips` that select based on service ID as opposed to primary key.

Reply:
I always used [index_match_object](https://erlang.org/doc/man/mnesia.html#index_match_object-4) or [index_read](https://erlang.org/doc/man/mnesia.html#index_read-3) when using the secondary indicies.